core/matrix/csr_kernels.hpp

```diff
@@ -216,6 +216,17 @@ namespace kernels {
                    const matrix::Dense<ValueType>* alpha,       \
                    matrix::Csr<ValueType, IndexType>* to_scale)
 
+#define GKO_DECLARE_CSR_CHECK_DIAGONAL_ENTRIES_EXIST(ValueType, IndexType) \
+    void check_diagonal_entries_exist(                                     \
+        std::shared_ptr<const DefaultExecutor> exec,                       \
+        const matrix::Csr<ValueType, IndexType>* mtx, bool* has_all_diags)
```

Comment:
nit: using a reference makes it clearer that this is a host value, compared to device data that is usually passed via raw pointers.
```suggestion
        const matrix::Csr<ValueType, IndexType>* mtx, bool& has_all_diags)
```

Reply:
This is consistent with `is_sorted_by_column_index`, `calculate_total_cols` etc. though.

Reply:
passing a pointer for a single out value is a C-ism that we don't need to continue doing. If we use pointers for device data and references for host data, the interface becomes more consistent.